Sources/SourceControl/GitRepository.swift

```diff
@@ -204,13 +204,21 @@ public struct GitRepositoryProvider: RepositoryProvider, Cancellable {
     }
 
     public func isValidDirectory(_ directory: Basics.AbsolutePath) throws -> Bool {
-        let result = try self.git.run(["-C", directory.pathString, "rev-parse", "--git-dir"])
-        return result == ".git" || result == "." || result == directory.pathString
+        do {
+            let result = try self.git.run(["-C", directory.pathString, "rev-parse", "--git-dir"])
+            return result == ".git" || result == "." || result == directory.pathString
+        } catch _ as GitShellError {
+            return false
+        }
```

Comment:
This means the error is now completely ignored, where exactly is its description printed then? This might also break  assumptions down the call stack about control flow, which previously relied on throwing this error.

Reply:
Right.

As I read it, there are two ways to interact with git repo:
 - low-level [GitHelper.run](https://github.com/swiftlang/swift-package-manager/blob/c237352345ff9deca691cc5c5e2e7b1cb216ae48/Sources/SourceControl/GitRepository.swift#L45) which doesn't have a context of what repo it's called for. It throws private exception [GitShellError](https://github.com/swiftlang/swift-package-manager/blob/c237352345ff9deca691cc5c5e2e7b1cb216ae48/Sources/SourceControl/GitRepository.swift#L1164).
 - high-level [GitRepository.callGit](https://github.com/swiftlang/swift-package-manager/blob/c237352345ff9deca691cc5c5e2e7b1cb216ae48/Sources/SourceControl/GitRepository.swift#L100) which does have a context of what repo it's called for and what human-readable message should be. It throws public exception [GitCloneError](https://github.com/swiftlang/swift-package-manager/blob/c237352345ff9deca691cc5c5e2e7b1cb216ae48/Sources/SourceControl/GitRepository.swift#L1200).

`isValidDirectory` uses the low-level, so it doesn't carry context of what was the repo and what should be the human-readable message. Its `GitShellError` also not caught anywhere further in the control flow. So if `isValidDirectory` raises exception, that's fatal for any Swift Package Manager operation.

For example, Swift Package Manager fails with this error if cache doesn't exist and there's no network connection:
```
error: GitShellError:     fatal: cannot change to '/Users/yy/Library/Caches/org.swift.swiftpm/repositories/postgres-kit-e4358808': No such file or directory
```

The error is accurate, but it's not helpful, because it doesn't tell what's wrong and how to fix that. Moreover, the raised exception is private, so there's no way of catching it gracefully. So changing the flow is absolutely the point.


```
error: Failed to clone repository https://github.com/vapor/postgres-kit.git:
    Cloning into bare repository '/Users/yy/work/pub/tst/testpackage/.build/repositories/postgres-kit-e4358808'...
    fatal: unable to access 'https://github.com/vapor/postgres-kit.git/': Could not resolve host: github.com
```

Where that message is coming from: for that specific use-case (offline and no pre-populated cache) there's [already code](https://github.com/swiftlang/swift-package-manager/blob/c237352345ff9deca691cc5c5e2e7b1cb216ae48/Sources/SourceControl/RepositoryManager.swift#L363  ) which checks offline, checks the cache, attempts to fetch, etc.

For the places where `isValidDirectory` is used, above is one of them,

 - no GitShellError exceptions are caught, because it's private
 - the flow would be the same for exception as well as non-valid dir

I did due diligence for "a control flow, which previously relied on throwing this error" at least within Swift Package Manager codebase, I believe it's not a huge factor for this change. I understand hesitation on changing behavior for public API, but the promise is ["returns true if the directory is valid git location"](https://github.com/swiftlang/swift-package-manager/blob/c237352345ff9deca691cc5c5e2e7b1cb216ae48/Sources/SourceControl/RepositoryManager.swift#L450), and I believe hiding private exception from public interface is within the promise.

Reply:
We still want it to throw the error in case there are other utilities using this API and want to display the error in their own way. All I thought this change needed to do was to make sure stderr was captured in the description which it wasn't previously and which you did fix below.

i.e. I agree with Max. :)

Reply:
:/ oh well

Can we make the `GitShelError` exception public then? Since the expectation that somebody may want to catch it and show in their own way outside of SourceControl module?

Reply:
I'd be OK with that.

Reply:
- change in the public behavior of the API is reverted in 2525234
 - the error struct is now public in 301fa08
 - description now includes the git command itself 18dea2c